[Docs] macOS Dev environment installation (#25204)
PX4/PX4-Autopilot: docs/en/dev_setup/dev_env_mac.md

# macOS Development Environment

The following instructions set up a PX4 development environment for macOS.
This environment can be used to build PX4 for:

- Pixhawk and other NuttX-based hardware
- [Gazebo Simulation](../sim_gazebo_gz/index.md) (Gazebo Harmonic)

:::tip
This setup is supported by the PX4 dev team.
To build other targets you will need to use a [different OS](../dev_setup/dev_env.md#supported-targets) or an [unsupported development environment](../advanced/community_supported_dev_env.md).
:::

## Video Guide

<lite-youtube videoid="tMbMGiMs1cQ" title="Setting up your PX4 development environment on macOS"/>

## Development Environment Setup

### Prerequisites

1. Install Homebrew by following these [installation instructions](https://brew.sh).

2. Enable more open files by appending the following line to the `~/.bashrc` (or `~/.zshrc` for zsh)  file or creating it if necessary:

   ```sh
   echo ulimit -S -n 2048 >> ~/.bashrc
   ```

   ::: info
   If you don't do this, the build toolchain may report the error: `"LD: too many open files"`
   :::

3. Enforce Python 3 by appending the following lines to `~/.bashrc`(or `~/.zshrc` for zsh):

   ```sh

>    # Point pip3 to macOS system python 3 pip

Reviewer: Since we are rewriting most of the docs already, why not update this section too:

I think this part is misguided, although it might work for some users, it depends on the base os level Python3 existing and the user not having installed a later version of Python3 via package manager or otherwise.

I suggest we say "Ensure Python 3 is used by default, since some scripts found in PX4 require python3 and pip3 executables in your path.

Then afterwards we can have a TIP that gives users a recommended way to install pyhon 3, which should say something like this

```
::: tip
If you need help installing an updated version of Python 3, we recommend the [pyenv project](https://github.com/pyenv/pyenv?tab=readme-ov-file), which gives you utmost flexibility by allowing you to set a global and local Python version at the per-directory level.
```

Reviewer 2: Before `brew install px4-dev px4-sim`:

```bash
mch@mchdexuniji ~ % which python3
/usr/bin/python3
mch@mchdexuniji ~ % which pip
pip not found
mch@mchdexuniji ~ % which pip3
pip3: aliased to /usr/bin/pip3

```
After running `brew install px4-dev px4-sim`:
```bash
mch@mchdexuniji ~ % which python3
/opt/homebrew/bin/python3
mch@mchdexuniji ~ % which pip3   
pip3: aliased to /usr/bin/pip3
mch@mchdexuniji ~ % brew list
==> Formulae
ant			gettext			lzo
arm-gcc-bin@13		giflib			m4
asio@1.10.8		glib			mpdecimal
astyle			graphite2		ncurses
autoconf		harfbuzz		ninja
automake		hiredis			openjdk
bash-completion		icu4c@77		openssl@3
blake3			jpeg-turbo		pcre2
boost@1.85		kconfig-frontends	pixman
ca-certificates		libpng			pkgconf
cairo			libtiff			px4-dev
ccache			libtommath		python-tk@3.13
cmake			libtool			python@3.13
expat			libunistring		readline
fastcdr			libx11			sqlite
fastdds			libxau			tcl-tk
flock			libxcb			tinyxml2
fmt			libxdmcp		xorgproto
fontconfig		libxext			xxhash
foonathan-memory	libxrender		xz
freetype		little-cms2		zstd
genromfs		lz4

==> Casks
mch@mchdexuniji ~ % 
```

   alias pip3=/usr/bin/pip3
   ```

### Install Development Tools

1. Install the PX4 development toolchain and simulation environment:

   ```sh
   brew tap PX4/px4
   brew install px4-dev px4-sim
   ```

2. **Important:** Link the ARM cross-compiler manually (required step):

   ```sh
   brew link --overwrite --force arm-gcc-bin@13
   ```

   ::: warning
   Homebrew does not link versioned formulae by default, so this manual linking step is required for the ARM cross-compiler (v13) to function properly.
   If you have other versions of `arm-none-eabi-gcc` installed, this may override them.
   You can unlink it manually later with `brew unlink arm-gcc-bin@13` if needed.
   :::

3. Install the required Python packages:

   ```sh
   # Install required packages using pip3
   python3 -m pip install --user pyserial empty toml numpy pandas jinja2 pyyaml pyros-genmsg packaging kconfiglib future jsonschema
   # If this fails with a permissions error, your Python install is in a system path - use this command instead:

>    sudo -H python3 -m pip install --user pyserial empty toml numpy pandas jinja2 pyyaml pyros-genmsg packaging kconfiglib future jsonschema

Reviewer: @farhangnaderi can you please double check that the packages are not part of `Tools/setup/requirements.txt` if you find anything missing can you please submit a pull request to add them?

Replace this section with
```
# Install required packages using pip3
pip3 install -r Tools/setup/requirements.txt
# if this fails with a permissions error, try the following
sudo -H pip3 install -r Tools/setup/requirements.txt
```

# macOS Development Environment

The following instructions set up a PX4 development environment for macOS.
This environment can be used to build PX4 for:

- Pixhawk and other NuttX-based hardware
- [Gazebo Simulation](../sim_gazebo_gz/index.md) (Gazebo Harmonic)

:::tip
This setup is supported by the PX4 dev team.
To build other targets you will need to use a [different OS](../dev_setup/dev_env.md#supported-targets) or an [unsupported development environment](../advanced/community_supported_dev_env.md).
:::

## Video Guide

<lite-youtube videoid="tMbMGiMs1cQ" title="Setting up your PX4 development environment on macOS"/>

## Development Environment Setup

### Prerequisites

1. Install Homebrew by following these [installation instructions](https://brew.sh).

> 2. Enable more open files by appending the following line to the `~/.zshenv` file (creating it if necessary):

Reviewer 3: I don't use .zshenv. Is this the default thing that everyone uses for all shells on macos?

   ::: tip
   If you need help installing an updated version of Python 3, we recommend the [pyenv project](https://github.com/pyenv/pyenv?tab=readme-ov-file), which gives you utmost flexibility by allowing you to set a global and local Python version at the per-directory level.
   :::

4. Point pip3 to macOS system python 3 pip:

   ```sh
   alias pip3=/usr/bin/pip3
   ```

### Install Development Tools

1. Install the PX4 development toolchain and simulation environment:

   ```sh
   brew tap PX4/px4
   brew install px4-dev px4-sim
   ```

2. Download PX4 Source Code:

   ```sh
   git clone https://github.com/PX4/PX4-Autopilot.git --recursive
   ```

3. Run the [`macos.sh` setup script](https://github.com/PX4/PX4-Autopilot/blob/main/Tools/setup/macos.sh):

   ```sh
   cd PX4-Autopilot
   ./Tools/setup/macos.sh
   ```

   The script automatically installs all development dependencies (homebrew packages, python libraries, and so on).

## Gazebo Simulation

Gazebo Harmonic simulation support is included with the `px4-sim` formula installed above.

> The simulation environment should be ready to use after the installation completes.

Reviewer 3: Is there a way to call macos.sh such that you don't install simulation tools?

This environment can be used to build PX4 for:

- Pixhawk and other NuttX-based hardware
- [Gazebo Simulation](../sim_gazebo_gz/index.md) (Gazebo Harmonic)

::: tip
This setup is supported by the PX4 dev team.
To build for [other targets](../dev_setup/dev_env.md#supported-targets) you will need to use a [different OS](../dev_setup/dev_env.md#supported-targets) or an [unsupported development environment](../advanced/community_supported_dev_env.md).
:::

<!-- ## Video Guide

> **Note:** The video guide previously available here is now outdated. Please follow the updated instructions below for the latest setup process. If you are referencing the YouTube video, be aware that some steps may have changed. -->

## Development Environment Setup

### Prerequisites

1. Install Homebrew by following these [installation instructions](https://brew.sh).

> 2. Enable more open files by appending the following line to the `~/.bashrc` (or `~/.zshrc` for zsh)  file or creating it if necessary:

Reviewer 3: Not quite right:

```suggestion
2. Enable more files to be opened in the shell process by appending `echo ulimit -S -n 2048` to your `~/.bashrc` file (or `~/.zshrc` for zsh).
   To append the line using the terminal enter:
```

- Pixhawk and other NuttX-based hardware
- [Gazebo Simulation](../sim_gazebo_gz/index.md) (Gazebo Harmonic)

::: tip
This setup is supported by the PX4 dev team.
To build for [other targets](../dev_setup/dev_env.md#supported-targets) you will need to use a [different OS](../dev_setup/dev_env.md#supported-targets) or an [unsupported development environment](../advanced/community_supported_dev_env.md).
:::

<!-- ## Video Guide

> **Note:** The video guide previously available here is now outdated. Please follow the updated instructions below for the latest setup process. If you are referencing the YouTube video, be aware that some steps may have changed. -->

## Development Environment Setup

### Prerequisites

1. Install Homebrew by following these [installation instructions](https://brew.sh).

2. Enable more files to be opened in the shell process by appending `echo ulimit -S -n 2048` to your `~/.bashrc` file (or `~/.zshrc` for zsh).
   To append the line using the terminal enter:

   ```sh
   echo ulimit -S -n 2048 >> ~/.bashrc
   ```

   ::: info
   If you don't do this, the build toolchain may report the error: `"LD: too many open files"`
   :::

> 3. Enforce Python 3 by appending the following lines to `~/.bashrc`(or `~/.zshrc` for zsh):

Reviewer 3: What lines do you wish to append?

   Ensure Python 3 is used by default, as some PX4 scripts require the `python3` and `pip3` executables to be available in your system `PATH`.

   ::: tip
   If you need help installing an updated version of Python 3, we recommend the [pyenv project](https://github.com/pyenv/pyenv?tab=readme-ov-file), which gives you utmost flexibility by allowing you to set a global and local Python version at the per-directory level.
   :::

4. Point pip3 to macOS system python 3 pip

   ```sh
   alias pip3=/usr/bin/pip3
   ```

### Install Development Tools

1. Install the PX4 development toolchain and simulation environment:

   ```sh
   brew tap PX4/px4
   brew install px4-dev px4-sim
   ```

> 2. ### Download PX4 Source Code

Reviewer 3: ```suggestion
2. Download PX4 Source Code
```

```sh
git clone https://github.com/PX4/PX4-Autopilot.git --recursive

> ```

Reviewer 3: ````suggestion
   ```sh
   git clone https://github.com/PX4/PX4-Autopilot.git --recursive
   ```
````

4. Prerequisites setup

>    Use the [`macos.sh` setup script](https://github.com/PX4/PX4-Autopilot/blob/main/Tools/setup/macos.sh) to automatically install all development dependencies:

Reviewer 3: ```suggestion
   Run the [`macos.sh` setup script](https://github.com/PX4/PX4-Autopilot/blob/main/Tools/setup/macos.sh) to automatically install all development dependencies (homebrew packages, python libraries, etc.):
```

   ```sh
   git clone https://github.com/PX4/PX4-Autopilot.git --recursive   cd PX4-Autopilot

>    ./Tools/setup/macos.sh

Reviewer 3: @farhangnaderi there are a lot of very basic errors here. Did you build the docs and look at them?

```suggestion
   cd PX4-Autopilot
   ./Tools/setup/macos.sh
```

   ```

>    This script will install all required Homebrew packages, Python dependencies, and perform the necessary setup steps.

Reviewer 3: ```suggestion
```

### Environment Setup

> :::details Apple Silicon Macbook users!

Reviewer 3: Does this means that the standard setup works for all types of silicon?

Reviewer: yes it works for all!

   sudo -H python3 -m pip install --user pyserial empty toml numpy pandas jinja2 pyyaml pyros-genmsg packaging kconfiglib future jsonschema
   ```

## Gazebo Simulation

To setup the environment for [Gazebo](../sim_gazebo_gz/index.md) simulation with Gazebo Harmonic:

1. Install Gazebo Harmonic and required dependencies:

   ```sh
   brew install gz-harmonic

>    brew install qt@5 gstreamer gst-plugins-base gst-plugins-good gst-plugins-bad gst-plugins-ugly gst-libav

Reviewer: does the gz-harmonic formula not install its dependencies?

Reviewer 3: Why are we doing this if `./PX4-Autopilot/Tools/setup/macos.sh --sim-tools` sets everything up?

   ```

> 1. Configure Qt5 and library paths by adding the following to your `~/.zshrc` file:

Reviewer: We need to reword this. You're not configuring Qt5; we're ensuring it's accessible to the default shell, which can be anything. The majority of users will be using `~/.bashrc`; let's default to this. Any Zsh users will know what to do from here.

The homebrew Qt5 formula gives you a very verbose explanation of what needs to happen after installation. We need to make it clear this isn't specific to PX4.

   ```sh
   # Qt5 configuration
   export Qt5_DIR=/opt/homebrew/opt/qt@5/lib/cmake/Qt5
   export PATH="/opt/homebrew/opt/qt@5/bin:$PATH"
   export CMAKE_PREFIX_PATH="/opt/homebrew/opt/qt@5:$CMAKE_PREFIX_PATH"

   # Library paths for GStreamer and other dependencies
   export PKG_CONFIG_PATH="/opt/homebrew/lib/pkgconfig:$PKG_CONFIG_PATH"
   export LIBRARY_PATH="/opt/homebrew/lib:$LIBRARY_PATH"
   export DYLD_LIBRARY_PATH="/opt/homebrew/lib:$DYLD_LIBRARY_PATH"
   ```

> 1. Link Qt5 and reload your shell configuration:

Reviewer: I recall that we had this discussion before. Although this can address a few initial issues, it's essentially a "magic" solution that isn't fully explained, and we shouldn't recommend it without also explaining its necessity. Let's find a way to be more explicit about it.

   ```sh
   brew unlink qt@5 && brew link --force qt@5
   source ~/.zshrc
   ```

1. Verify Qt5 installation:

   ```sh

>    ls -la /opt/homebrew/lib/cmake/Qt5Core/

Reviewer: Why is this step necessary? You already installed Qt5 twice above; this seems redundant. What's the issue we are trying to solve?

   ```

1. Run the macOS setup script: `PX4-Autopilot/Tools/setup/macos.sh`
   The easiest way to do this is to clone the PX4 source, and then run the script from the directory, as shown:

   ```sh
   git clone https://github.com/PX4/PX4-Autopilot.git --recursive
   cd PX4-Autopilot/Tools/setup
   sh macos.sh
   ```

## Unsupported Simulation Options

> ### Gazebo Classic

Author: same question here.
